### ISSUE TITLE
MirrorStrategy() fills up memory of GPU that is not selected for training in TF 2.0.

### ISSUE THREAD

**Issue by rishabhsahrawat** [tensorflow/tensorflow#31410]
I am using `tf.distribute.MirroredStrategy()` in order to train a `tf.keras` model. I have 2 T4 GPUs from Google available. I want  to train my model consisting LSTM layer on only one GPU i.e. `/gpu:1`. So, in order to select it, I define `devices = ["/gpu:1"]` inside `MirrorStrategy()` as suggested [here](https://www.tensorflow.org/guide/distribute_strategy#mirroredstrategy), now when I run the file for training, the memory of first GPU(/gpu:0) is also filling up completely however, during training only GPU (/gpu:1) is being utilized completely. My question is why is this happening since I only want to utilize GPU (/gpu:1)? In this case, first GPU is useless.
However, if I select the first GPU(/gpu:0) for training then it uses only 112MB of memory of the second GPU(/gpu:1).
```
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla T4            On   | 00000000:00:04.0 Off |                    0 |
| N/A   60C    P0    28W /  70W |  14449MiB / 15079MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla T4            On   | 00000000:00:05.0 Off |                    0 |
| N/A   75C    P0    74W /  70W |  14517MiB / 15079MiB |     95%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     20576      C   python                                     14439MiB |
|    1     20576      C   python                                     14507MiB |
+-----------------------------------------------------------------------------+

**Comment by srihari-humbarwadi** [comment]
@rishabhsahrawat are you executing any other op before you declare the distributed strategy ?
it would be easier to comment on this if you could provide a reproducible code snippet!

**Comment by rishabhsahrawat** [comment]
Hi @srihari-humbarwadi , thank you for your questions. No, I am not declaring any op as far as I know. I am only creating dataset for training and testing. 
Yes, it is still happening even if I define distribute strategy right after defining modules at the top of the file.
If you use the code from this [tutorial](https://www.tensorflow.org/beta/tutorials/load_data/text#split_the_dataset_into_text_and_train_batches) after defining the model inside the `MirrorStrategy()` like 
````
mirrored_strategy = tf.distribute.MirroredStrategy(devices = ['/gpu:1'])# for using second GPU

with mirrored_strategy.scope():
  model = tf.keras.Sequential()
  model.add(tf.keras.layers.Embedding(17179, 64))
  model.add(tf.keras.layers.Bidirectional(tf.keras.layers.LSTM(64, activation= 'tanh', recurrent_activation= 'sigmoid', recurrent_dropout = 0, unroll = False, use_bias= True)))
  # One or more dense layers.
  # Edit the list in the `for` line to experiment with layer sizes.
  for units in [64, 64]:
    model.add(tf.keras.layers.Dense(units, activation='relu'))

  # Output layer. The first argument is the number of labels.
  model.add(tf.keras.layers.Dense(3, activation='softmax'))
  model.compile(optimizer='adam',
                loss='sparse_categorical_crossentropy',
                metrics=['accuracy'])
````
, you will be able to see this behaviour. You can do it on your own just to be sure otherwise, here is the full script.
````
import tensorflow as tf
import tensorflow_datasets as tfds
import os
DIRECTORY_URL = 'https://storage.googleapis.com/download.tensorflow.org/data/illiad/'
FILE_NAMES = ['cowper.txt', 'derby.txt', 'butler.txt']

for name in FILE_NAMES:
  text_dir = tf.keras.utils.get_file(name, origin=DIRECTORY_URL+name)
  
parent_dir = os.path.dirname(text_dir)

print parent_dir
def labeler(example, index):
  return example, tf.cast(index, tf.int64)  

labeled_data_sets = []

for i, file_name in enumerate(FILE_NAMES):
  lines_dataset = tf.data.TextLineDataset(os.path.join(parent_dir, file_name))
  labeled_dataset = lines_dataset.map(lambda ex: labeler(ex, i))
  labeled_data_sets.append(labeled_dataset)

BUFFER_SIZE = 50000
BATCH_SIZE = 64
TAKE_SIZE = 5000
all_labeled_data = labeled_data_sets[0]
for labeled_dataset in labeled_data_sets[1:]:
  all_labeled_data = all_labeled_data.concatenate(labeled_dataset)
  
all_labeled_data = all_labeled_data.shuffle(
    BUFFER_SIZE, reshuffle_each_iteration=False)

for ex in all_labeled_data.take(1):
  print(ex)


tokenizer = tfds.features.text.Tokenizer()

vocabulary_set = set()
for text_tensor, _ in all_labeled_data:
  some_tokens = tokenizer.tokenize(text_tensor.numpy())
  vocabulary_set.update(some_tokens)

vocab_size = len(vocabulary_set)
print 'coab size: ',vocab_size

encoder = tfds.features.text.TokenTextEncoder(vocabulary_set)
example_text = next(iter(all_labeled_data))[0].numpy()
print(example_text)
encoded_example = encoder.encode(example_text)
print(encoded_example)

def encode(text_tensor, label):
  encoded_text = encoder.encode(text_tensor.numpy())
  #encoded_text = encoded_text[:10]
  return encoded_text, label

def encode_map_fn(text, label):
  return tf.py_function(encode, inp=[text, label], Tout=(tf.int64, tf.int64))

all_encoded_data = all_labeled_data.map(encode_map_fn)

train_data = all_encoded_data.skip(TAKE_SIZE).shuffle(BUFFER_SIZE)
train_data = train_data.padded_batch(BATCH_SIZE, padded_shapes=([-1],[]))
#train_data= train_data.batch(64)
test_data = all_encoded_data.take(TAKE_SIZE)

test_data = test_data.padded_batch(BATCH_SIZE, padded_shapes=([-1],[]), drop_remainder = False)
#test_data = test_data.batch(64)
sample_text, sample_labels = next(iter(test_data))
print '**Test Samples**'
print sample_text[0], sample_labels[0]

vocab_size += 1
###Distributed Strategy###
mirrored_strategy = tf.distribute.MirroredStrategy(devices = ['/gpu:1'])
with mirrored_strategy.scope():
  model = tf.keras.Sequential()
  model.add(tf.keras.layers.Embedding(vocab_size, 64))
  model.add(tf.keras.layers.Bidirectional(tf.keras.layers.LSTM(64, activation= 'tanh', recurrent_activation= 'sigmoid', recurrent_dropout = 0, unroll = False, use_bias= True)))
  # One or more dense layers.
  # Edit the list in the `for` line to experiment with layer sizes.
  for units in [64, 64]:
    model.add(tf.keras.layers.Dense(units, activation='relu'))

  # Output layer. The first argument is the number of labels.
  model.add(tf.keras.layers.Dense(3, activation='softmax'))
  model.compile(optimizer='adam',
                loss='sparse_categorical_crossentropy',
                metrics=['accuracy'])
###Strategy  end###
steps = 5000/64

print steps

model.fit(train_data, epochs=10, validation_steps = steps, validation_data=test_data)
````
Thank you,
Rishabh Sahrawat

**Comment by srihari-humbarwadi** [comment]
The distribute strategy declaration should be done before you use any tf operations (like tf.data, tf.keras etc). Please post what happens when you declare this at the start of your code.
```
mirrored_strategy = tf.distribute.MirroredStrategy(devices = ['/gpu:1'])
```

**Comment by rishabhsahrawat** [comment]
I already mentioned it in my last comment. But another try is here.

> Yes, it is still happening even if I define distribute strategy right after defining modules at the top of the file.

````
import tensorflow as tf
import tensorflow_datasets as tfds
import os
mirrored_strategy = tf.distribute.MirroredStrategy(devices = ['/gpu:1'])
--rest is same like script in last comment (of course after removing the above line from it)--
````
The result is still the same, it is using both GPUs for memory but utilising second.
````
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla T4            On   | 00000000:00:04.0 Off |                    0 |
| N/A   44C    P0    26W /  70W |  14449MiB / 15079MiB |      2%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla T4            On   | 00000000:00:05.0 Off |                    0 |
| N/A   43C    P0    31W /  70W |  14827MiB / 15079MiB |     40%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     16200      C   python                                     14439MiB |
|    1     16200      C   python                                     14817MiB |
+-----------------------------------------------------------------------------+
````

**Comment by srihari-humbarwadi** [comment]
Just add the tf.data ops under the distribute strategy scope

**Comment by rishabhsahrawat** [comment]
so you mean like this?
````
import tensorflow as tf
import tensorflow_datasets as tfds
import os
mirrored_strategy = tf.distribute.MirroredStrategy(devices = ['/gpu:0'])
DIRECTORY_URL = 'https://storage.googleapis.com/download.tensorflow.org/data/illiad/'
FILE_NAMES = ['cowper.txt', 'derby.txt', 'butler.txt']

for name in FILE_NAMES:
  text_dir = tf.keras.utils.get_file(name, origin=DIRECTORY_URL+name)
  
parent_dir = os.path.dirname(text_dir)
print parent_dir

#CHANGES START HERE ON
with mirrored_strategy.scope():
  def labeler(example, index):
    return example, tf.cast(index, tf.int64)  

  labeled_data_sets = []

  for i, file_name in enumerate(FILE_NAMES):
    lines_dataset = tf.data.TextLineDataset(os.path.join(parent_dir, file_name))
    labeled_dataset = lines_dataset.map(lambda ex: labeler(ex, i))
    labeled_data_sets.append(labeled_dataset)

  BUFFER_SIZE = 50000
  BATCH_SIZE = 64
  TAKE_SIZE = 5000
  all_labeled_data = labeled_data_sets[0]
  for labeled_dataset in labeled_data_sets[1:]:
    all_labeled_data = all_labeled_data.concatenate(labeled_dataset)
    
  all_labeled_data = all_labeled_data.shuffle(
      BUFFER_SIZE, reshuffle_each_iteration=False)

  for ex in all_labeled_data.take(1):
    print(ex)


  tokenizer = tfds.features.text.Tokenizer()

  vocabulary_set = set()
  for text_tensor, _ in all_labeled_data:
    some_tokens = tokenizer.tokenize(text_tensor.numpy())
    vocabulary_set.update(some_tokens)

  vocab_size = len(vocabulary_set)
  print 'coab size: ',vocab_size

  encoder = tfds.features.text.TokenTextEncoder(vocabulary_set)
  example_text = next(iter(all_labeled_data))[0].numpy()
  print(example_text)
  encoded_example = encoder.encode(example_text)
  print(encoded_example)

  def encode(text_tensor, label):
    encoded_text = encoder.encode(text_tensor.numpy())
    #encoded_text = encoded_text[:10]
    return encoded_text, label

  def encode_map_fn(text, label):
    return tf.py_function(encode, inp=[text, label], Tout=(tf.int64, tf.int64))

  all_encoded_data = all_labeled_data.map(encode_map_fn)

  train_data = all_encoded_data.skip(TAKE_SIZE).shuffle(BUFFER_SIZE)
  train_data = train_data.padded_batch(BATCH_SIZE, padded_shapes=([-1],[]))
  #train_data= train_data.batch(64)
  test_data = all_encoded_data.take(TAKE_SIZE)

  test_data = test_data.padded_batch(BATCH_SIZE, padded_shapes=([-1],[]), drop_remainder = False)
  #test_data = test_data.batch(64)
  sample_text, sample_labels = next(iter(test_data))
  print '**Test Samples**'
  print sample_text[0], sample_labels[0]

  vocab_size += 1

  model = tf.keras.Sequential()
  model.add(tf.keras.layers.Embedding(vocab_size, 64))
  model.add(tf.keras.layers.Bidirectional(tf.keras.layers.LSTM(64, activation= 'tanh', recurrent_activation= 'sigmoid', recurrent_dropout = 0, unroll = False, use_bias= True)))
  # One or more dense layers.
  # Edit the list in the `for` line to experiment with layer sizes.
  for units in [64, 64]:
    model.add(tf.keras.layers.Dense(units, activation='relu'))

  # Output layer. The first argument is the number of labels.
  model.add(tf.keras.layers.Dense(3, activation='softmax'))
  model.compile(optimizer='adam',
                loss='sparse_categorical_crossentropy',
                metrics=['accuracy'])
steps = 5000/64
print steps
model.fit(train_data, epochs=10, validation_steps = steps, validation_data=test_data)

**Comment by rishabhsahrawat** [comment]
Unfortunately, it still has the same behaviour.

**Comment by isaprykin** [comment]
I am surprised that setting the scope:gpu1 at the very top doesn't work as well.

Could you try running with the environmental variable CUDA_VISIBLE_DEVICES=1 (see more info http://stackoverflow.com/q/37893755), please?

**Comment by rishabhsahrawat** [comment]
Hi @isaprykin , thank you for your suggestion. I tried it and can confirm this works however one must not define `devices` parameter of the strategy, which also makes this parameter useless since the way to choose a GPU doesn't depend on `devices` parameter. Also I would request you to add this in the documentation.

**Comment by rishabhsahrawat** [comment]
I am facing a weird situation. When I train the same model on one GPU, it takes 45 secs to process 100 batches during first epoch, but when I use both GPUs for training, it takes 61 secs for the same however it should train faster on more number fo GPUs. Do you guys think there is a reason for that?

**Comment by isaprykin** [comment]
Are both GPUs the same?

Profiling tutorial is here: https://www.tensorflow.org/tensorboard/r2/tensorboard_profiling_keras

**Comment by rishabhsahrawat** [comment]
Yes, both GPUs are same Tesla T4s from Google. I also tried using 2 K80s but it throws OOM errors then. 
I already tried using `tf.data.prefetch()` but it doesn't show any big improvement, with the use of more number fo GPUs it gets slower in training.
